Proposal: enhance pluggable policy

deads2k opened the pull request:
This describes modifications to the current authorization APIs to make policy more powerful and easier to extend.  This document aligns with OpenShift authorization.

@erictune ptal

@liggitt @smarterclayton @derekwaynecarr fyi

deads2k committed:
enhance pluggable policy

deads2k commented:
/cc @preillyme @uruddarraju

smarterclayton commented on docs/design/enhance-pluggable-policy.md:
```diff
@@ -0,0 +1,175 @@
+
+  // GetAllowedSubjects takes a Context (for namespace and traceability) and Attributes to determine which users and 
+  // groups are allowed to perform the described action in the namespace.    This API enables the ResourceBasedReview requests below
+  GetAllowedSubjects(ctx api.Context, a Attributes) (users util.StringSet, groups util.StringSet, evaluationError error)
```
Maybe split this into a second interface which is optionally provided.

deads2k replied:
> Maybe split this into a second interface which is optionally provided.

`GetAllowedSubjects` is a pretty fundamental operation.  Without the ability to know who can perform an operation, proper auditing of authorization policy is an impossibility.  Why would we ever want to allow someone to create an authorizer that couldn't be audited?

smarterclayton replied:
Because you don't need audit?  Unless every call site needs both methods
(and they don't) I would organize these so that there are two fundamental
interfaces, and if at the highest level you require both, require both.

On Thu, Aug 6, 2015 at 9:30 AM, David Eads <notifications@github.com> wrote:

> In docs/design/enhance-pluggable-policy.md
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/12209#discussion_r36413537>
> :
>
> > +// OLD
> > +type Authorizer interface {
> > +  Authorize(a Attributes) error
> > +}
> > +```
> > +
> > +```
> > +// NEW
> > +type Authorizer interface {
> > +  // Authorize takes a Context (for namespace, user, and traceability) and Attributes to make a policy determination.
> > +  // reason is an optional return value that can describe why a policy decision was made.
> > +  Authorize(ctx api.Context, a Attributes) (allowed bool, reason string, evaluationError error)
> > +
> > +  // GetAllowedSubjects takes a Context (for namespace and traceability) and Attributes to determine which users and
> > +  // groups are allowed to perform the described action in the namespace.    This API enables the ResourceBasedReview requests below
> > +  GetAllowedSubjects(ctx api.Context, a Attributes) (users util.StringSet, groups util.StringSet, evaluationError error)
>
> Maybe split this into a second interface which is optionally provided.
>
> GetAllowedSubjects is a pretty fundamental operation. Without the ability
> to know who can perform an operation, proper auditing of authorization
> policy is an impossibility. Why would we ever want to allow someone to
> create an authorizer that couldn't be audited?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/12209/files#r36413537>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

deads2k replied:
> Because you don't need audit?  Unless every call site needs both methods
> (and they don't) I would organize these so that there are two fundamental
> interfaces, and if at the highest level you require both, require both.

So would we make `ResourceAccessReview` optional?  It's essential for doing things like "which namespaces/projects can I see".

smarterclayton replied:
I don't think ResourceAccessReview is required for all deployments or
authorizers.  It is definitely useful, but not all consumers may choose to
invert the model (or even support inverting the model).

On Fri, Aug 7, 2015 at 11:38 AM, David Eads <notifications@github.com>
wrote:

> In docs/design/enhance-pluggable-policy.md
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/12209#discussion_r36531053>
> :
>
> > +// OLD
> > +type Authorizer interface {
> > +  Authorize(a Attributes) error
> > +}
> > +```
> > +
> > +```
> > +// NEW
> > +type Authorizer interface {
> > +  // Authorize takes a Context (for namespace, user, and traceability) and Attributes to make a policy determination.
> > +  // reason is an optional return value that can describe why a policy decision was made.
> > +  Authorize(ctx api.Context, a Attributes) (allowed bool, reason string, evaluationError error)
> > +
> > +  // GetAllowedSubjects takes a Context (for namespace and traceability) and Attributes to determine which users and
> > +  // groups are allowed to perform the described action in the namespace.    This API enables the ResourceBasedReview requests below
> > +  GetAllowedSubjects(ctx api.Context, a Attributes) (users util.StringSet, groups util.StringSet, evaluationError error)
>
> Because you don't need audit? Unless every call site needs both methods
> (and they don't) I would organize these so that there are two fundamental
> interfaces, and if at the highest level you require both, require both.
>
> So would we make ResourceAccessReview optional? It's essential for doing
> things like "which namespaces/projects can I see".
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/12209/files#r36531053>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

smarterclayton commented on docs/design/enhance-pluggable-policy.md:
```diff
@@ -0,0 +1,175 @@
+// input
+{
+  "kind": "ResourceAccessReview",
+  "apiVersion": "v1beta3",
```
Switch examples to v1

smarterclayton commented on docs/design/enhance-pluggable-policy.md:
````diff
@@ -0,0 +1,175 @@
+```
+
+### SubjectAccessReview
+`/api/{version}/ns/{namespace}/subjectAccessReview` - This API answers the question: can a user or group (use authenticated user if none is specified) perform a given action.  Given the Authorizer interface (proposed or existing), this endpoint can be implemented generically against any Authorizer by creating the correct Attributes and making an .Authorize() call.
````
Propose this under a new API group - policy or authorization or similar.

smarterclayton commented on docs/design/enhance-pluggable-policy.md:
```diff
@@ -0,0 +1,175 @@
+### SubjectAccessReview
+`/api/{version}/ns/{namespace}/subjectAccessReview` - This API answers the question: can a user or group (use authenticated user if none is specified) perform a given action.  Given the Authorizer interface (proposed or existing), this endpoint can be implemented generically against any Authorizer by creating the correct Attributes and making an .Authorize() call.
+
+SubjectAccessReview is runtime.Object with associated RESTStorage that only accepts creates.  The caller POSTs a SubjectAccessReview to this URL and he gets a SubjectAccessReviewResponse back.  Here is an example of a call and its corresponding return.
```
We may want to make this a SubjectAccessResponse or unify the struct between request and response - action being one nested struct and response being the other (or just use status).

pmorie replied:
@ncdc @pweil-

On Tuesday, August 4, 2015, Clayton Coleman <notifications@github.com>
wrote:

> In docs/design/enhance-pluggable-policy.md
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/12209#discussion_r36258015>
> :
>
> > +// NEW
> > +type Authorizer interface {
> > +  // Authorize takes a Context (for namespace, user, and traceability) and Attributes to make a policy determination.
> > +  // reason is an optional return value that can describe why a policy decision was made.
> > +  Authorize(ctx api.Context, a Attributes) (allowed bool, reason string, evaluationError error)
> > +
> > +  // GetAllowedSubjects takes a Context (for namespace and traceability) and Attributes to determine which users and
> > +  // groups are allowed to perform the described action in the namespace.    This API enables the ResourceBasedReview requests below
> > +  GetAllowedSubjects(ctx api.Context, a Attributes) (users util.StringSet, groups util.StringSet, evaluationError error)
> > +}
> > +```
> > +
> > +### SubjectAccessReview
> > +`/api/{version}/ns/{namespace}/subjectAccessReview` - This API answers the question: can a user or group (use authenticated user if none is specified) perform a given action.  Given the Authorizer interface (proposed or existing), this endpoint can be implemented generically against any Authorizer by creating the correct Attributes and making an .Authorize() call.
> > +
> > +SubjectAccessReview is runtime.Object with associated RESTStorage that only accepts creates.  The caller POSTs a SubjectAccessReview to this URL and he gets a SubjectAccessReviewResponse back.  Here is an example of a call and its corresponding return.
>
> We may want to make this a SubjectAccessResponse or unify the struct
> between request and response - action being one nested struct and response
> being the other (or just use status).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/12209/files#r36258015>
> .
>

pmorie commented:
Nit: I think it's good to use backticks around code constructs

pmorie commented on docs/design/enhance-pluggable-policy.md:
```diff
@@ -0,0 +1,175 @@
+ 1.  Provide a way for disconnected pieces of the system to ask whether a user is permitted to take an action without running in process with the API Authorizer.  For instance, a proxy for exec calls could ask whether a user can run the exec they are requesting.
+ 1.  Provide a mechanism to discover who can perform a given action on a given resource.  This is useful for answering questions like, "who can create replication controllers in my namespace".
+
+This proposal adds to and extends existing API to so that authorizers may provide the functionality described above.  It does not attempt to describe how the policies themselves can be expressed, that is up the authorization plugins themselves.
```
extends the existing API

pmorie commented on docs/design/enhance-pluggable-policy.md:
```diff
@@ -0,0 +1,175 @@
+While trying to develop an authorization plugin for Kubernetes, we found a few places where API extensions would ease development and add power.  There are a few goals:
+ 1.  Provide an authorization plugin that can evaluate a .Authorize() call based on the full content of the request to RESTStorage.  This includes information like the full verb, the content of creates and updates, and the names of resources being acted upon.
+ 1.  Provide a mechanism for a user to determine whether or not he can take an action without attempting the action itself.
+ 1.  Provide a way for disconnected pieces of the system to ask whether a user is permitted to take an action without running in process with the API Authorizer.  For instance, a proxy for exec calls could ask whether a user can run the exec they are requesting.
```
Is this the same as the above bullet?

deads2k replied:
> Is this the same as the above bullet?

The one above allows in-process queries, the later one allow out-of-process queries.

derekwaynecarr replied:
this confused me as well, maybe say provide a REST API to ask whether a user is permitted to take an action.

pmorie commented on docs/design/enhance-pluggable-policy.md:
```diff
@@ -0,0 +1,175 @@
+### Request Attributes
+The existing authorizer.Attributes only has 5 attributes (user, groups, isReadOnly, kind, and namespace).  If we add more detailed verbs, content, and resource names, then Authorizer plugins will have the same level of information available to RESTStorage components in order to express more detailed policy.  The replacement excerpt is below.
+
+A request has 7 attributes that can be considered for authorization:
```
Is there an analog for `readOnly` in the new regime?

deads2k replied:
> Is there an analog for readOnly in the new regime?

readOnly is equivalent to allowing "get", "list", and "watch" verbs.

pmorie replied:
Ok -- I think it would be good to spell that out, since it's presented as 5 attributes versus 7 attributes, but it's not strictly additive.

deads2k replied:
note to self, add `nonResourceURL` similar to openshift to still allow full filtering of the entire mux.

smarterclayton replied:
Have we had any serious disagreement to this proposal so far?  I haven't
heard any....

On Sep 4, 2015, at 10:49 AM, David Eads <notifications@github.com> wrote:

In docs/design/enhance-pluggable-policy.md
<https://github.com/kubernetes/kubernetes/pull/12209#discussion_r38758883>:

> +While trying to develop an authorization plugin for Kubernetes, we found a few places where API extensions would ease development and add power.  There are a few goals:
> + 1.  Provide an authorization plugin that can evaluate a .Authorize() call based on the full content of the request to RESTStorage.  This includes information like the full verb, the content of creates and updates, and the names of resources being acted upon.
> + 1.  Provide a mechanism for a user to determine whether or not he can take an action without attempting the action itself.
> + 1.  Provide a way for disconnected pieces of the system to ask whether a user is permitted to take an action without running in process with the API Authorizer.  For instance, a proxy for exec calls could ask whether a user can run the exec they are requesting.
> + 1.  Provide a mechanism to discover who can perform a given action on a given resource.  This is useful for answering questions like, "who can create replication controllers in my namespace".
> +
> +This proposal adds to and extends existing API to so that authorizers may provide the functionality described above.  It does not attempt to describe how the policies themselves can be expressed, that is up the authorization plugins themselves.
> +
> +
> +## Enhancements to existing Authorization interfaces
> +The existing Authorization interfaces are described here: [authorization.md](../authorization.md).  A couple additions will allow the development of an Authorizer that matches based on different rules than the existing implementation.
> +
> +### Request Attributes
> +The existing authorizer.Attributes only has 5 attributes (user, groups, isReadOnly, kind, and namespace).  If we add more detailed verbs, content, and resource names, then Authorizer plugins will have the same level of information available to RESTStorage components in order to express more detailed policy.  The replacement excerpt is below.
> +
> +A request has 7 attributes that can be considered for authorization:

note to self, add nonResourceURL similar to openshift to still allow full
filtering of the entire mux.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/12209/files#r38758883>.

derekwaynecarr commented on docs/design/enhance-pluggable-policy.md:
```diff
@@ -0,0 +1,175 @@
+# Enhance Pluggable Policy
+While trying to develop an authorization plugin for Kubernetes, we found a few places where API extensions would ease development and add power.  There are a few goals:
+ 1.  Provide an authorization plugin that can evaluate a .Authorize() call based on the full content of the request to RESTStorage.  This includes information like the full verb, the content of creates and updates, and the names of resources being acted upon.
+ 1.  Provide a mechanism for a user to determine whether or not he can take an action without attempting the action itself.
```
clarify that this is in process of api-server.

derekwaynecarr commented on docs/design/enhance-pluggable-policy.md:
```diff
@@ -0,0 +1,175 @@
+ 1.  Provide an authorization plugin that can evaluate a .Authorize() call based on the full content of the request to RESTStorage.  This includes information like the full verb, the content of creates and updates, and the names of resources being acted upon.
+ 1.  Provide a mechanism for a user to determine whether or not he can take an action without attempting the action itself.
+ 1.  Provide a way for disconnected pieces of the system to ask whether a user is permitted to take an action without running in process with the API Authorizer.  For instance, a proxy for exec calls could ask whether a user can run the exec they are requesting.
+ 1.  Provide a mechanism to discover who can perform a given action on a given resource.  This is useful for answering questions like, "who can create replication controllers in my namespace".
```
clarify that this is a REST API?

deads2k replied:
> clarify that this is a REST API?

I've collapsed this down, deciding that one covers the other.  I haven't indicated that this is REST API, because if the `Authorizer` conforms to the interface, the REST API simply exposes this access without the provider having to do anything.

derekwaynecarr commented on docs/design/enhance-pluggable-policy.md:
```diff
@@ -0,0 +1,175 @@
+A request has 7 attributes that can be considered for authorization:
+  - user - the user-string which a user was authenticated as.  This is included in the Context.
+  - groups - the groups to which the user belongs.  This is included in the Context.
+  - verb - string describing the requestion action. Today we have: get, list, watch, create, update, and delete
```
nit: requesting action.  

is ``connect``` not an action?

deads2k replied:
> is connect` not an action?

Actually no.  I could see revisiting that, but we haven't special cased it yet.

derekwaynecarr commented on docs/design/enhance-pluggable-policy.md:
```diff
@@ -0,0 +1,175 @@
+    - applies only to the API endpoints, such as 
+        `/api/v1beta1/pods`.  For miscelaneous endpoints, like `/version`, the kind is the empty string.
+  - the namespace of the object being access, or the empty string if the endpoint does not support namespaced objects.  This is included in the Context.
+  - content - the runtime.Object being submitted. This only has content for creates and updates
```
maybe make clear that this is prior to any admission control or defaulting?

derekwaynecarr commented on docs/design/enhance-pluggable-policy.md:
```diff
@@ -0,0 +1,175 @@
+        `/api/v1beta1/pods`.  For miscelaneous endpoints, like `/version`, the kind is the empty string.
+  - the namespace of the object being access, or the empty string if the endpoint does not support namespaced objects.  This is included in the Context.
+  - content - the runtime.Object being submitted. This only has content for creates and updates
+  - resourceName - the name of the resource being gotten (get-ed) or deleted.
```
the name of the resource during a get, watch, update, or delete action.

deads2k replied:
> the name of the resource during a get, watch, update, or delete action

I thought watch was a special kind of list

derekwaynecarr commented on docs/design/enhance-pluggable-policy.md:
```diff
@@ -0,0 +1,175 @@
+  - resourceName - the name of the resource being gotten (get-ed) or deleted.
+
+### Authorizer Interface
+The existing Authorizer interface is very simple, but there isn't a way to provide details about allows, denies, or failures.  The extended detail is useful for UIs that want to describe why certain actions are disallowed.  Not all Authorizers will want to provide that information, but for those that do having that capability is useful.  In addition, adding a `GetAllowedSubjects` method that retu ...
```
nit: readability. "Not all Authorizers will want to provide that information, but for those that do, having that capability is useful."

pmorie replied:
Can anyone call `GetAllowedSubjects`, or do you have to be allowed? No pun intended.

deads2k replied:
> Can anyone call GetAllowedSubjects, or do you have to be allowed? No pun intended.

Anyone in process with direct access to the authorizer can call whatever they want.  Any REST API that exposes the `GetAllowedSubjects` call would be protected by the authorizer itself.

derekwaynecarr commented on docs/design/enhance-pluggable-policy.md:
```diff
@@ -0,0 +1,175 @@
+type Authorizer interface {
+  // Authorize takes a Context (for namespace, user, and traceability) and Attributes to make a policy determination.
+  // reason is an optional return value that can describe why a policy decision was made.
+  Authorize(ctx api.Context, a Attributes) (allowed bool, reason string, evaluationError error)
```
You did not make a goal explaining why ```reason``` is needed or what value it provides.

Maybe include some information on why the ```reason``` is useful, if it should be made visible to the end-user or logged only, etc?

pmorie replied:
@derekwaynecarr Isn't this addressed in the passage that describes UIs that want to display reason information?

deads2k commented:
First round comments done.  Items not addressed:

1. 

> Maybe split this into a second interface which is optionally provided.

`GetAllowedSubjects` is a pretty fundamental operation.  Without the ability to know who can perform an operation, proper auditing of authorization policy is an impossibility.  Why would we ever want to allow someone to create an authorizer that couldn't be audited? 

2.

> We may want to make this a SubjectAccessResponse or unify the struct between request and response - action being one nested struct and response being the other (or just use status).

We considered `spec`/`status` before, but it was rejected (https://github.com/GoogleCloudPlatform/kubernetes/pull/3730#discussion_r23418439).  I like it, but I'd like to hear from @erictune  before adding it back in.

smarterclayton commented:
On Thu, Aug 6, 2015 at 9:39 AM, David Eads <notifications@github.com> wrote:

> First round comments done. Items not addressed:
>
>    1.
>
> Maybe split this into a second interface which is optionally provided.
>
> GetAllowedSubjects is a pretty fundamental operation. Without the ability
> to know who can perform an operation, proper auditing of authorization
> policy is an impossibility. Why would we ever want to allow someone to
> create an authorizer that couldn't be audited?
>
> 2.
>
> We may want to make this a SubjectAccessResponse or unify the struct
> between request and response - action being one nested struct and response
> being the other (or just use status).
>
> We considered spec/status before, but it was rejected (#3730 (comment)
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/3730#discussion_r23418439>).
> I like it, but I'd like to hear from @erictune
> <https://github.com/erictune> before adding it back in.
>

It's worth discussing whether the world has changed since then - it's
easier to reason about one object (a POST that materializes the status of
the incoming request) than two in client code, and when I get a response I
like to also see what I requested.

erictune commented:
I'm on vacation this week (:house: :mouse:) and will take a look next week.

deads2k commented:
> Have we had any serious disagreement to this proposal so far?  I haven't

I haven't seen disagreement.  `AccessReview` and `AccessReviewResponse` versus spec/status is still an open question, but its not a significant implementation change either way.

smarterclayton commented:
Hrm... spec/status definitely fits with our other models.

liggitt commented:
yes... though when you "create" a review, you aren't actually persisting anything, so it feels little silly to echo back your request

smarterclayton commented:
You POST and get back an object that has more details.

smarterclayton commented:
There's very little about REST that requires persisting anything.  The
pattern is "noun" with four verbs - if you're POSTing a partial resource
filling it out with extra data is probably just as bizarre as returning a
completely different resource.  Since we have "status" pretty firmly
entrenched in the API, I don't see its use for the effective state of the
cluster (observed value) as being terribly different.

On Thu, Sep 10, 2015 at 11:15 AM, Kubernetes Bot <notifications@github.com>
wrote:

> GCE e2e build/test *passed* for commit 793ea6d
> <https://github.com/kubernetes/kubernetes/commit/793ea6d551c938eecede1fd6081a7f61db1c66df>
> .
>
>    - Build Log
>    <https://storage.cloud.google.com/kubernetes-jenkins/pr-logs/793ea6d551c938eecede1fd6081a7f61db1c66df/kubernetes-pull-build-test-e2e-gce/7987/build-log.txt>
>    - Test Artifacts
>    <https://console.developers.google.com/storage/browser/kubernetes-jenkins/pr-logs/793ea6d551c938eecede1fd6081a7f61db1c66df/kubernetes-pull-build-test-e2e-gce/7987/_artifacts/>
>    - Internal Jenkins Results
>    <http://goto.google.com/prkubekins/job/kubernetes-pull-build-test-e2e-gce//7987>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/12209#issuecomment-139276305>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

smarterclayton commented:
@erictune can you comment on this and review?  Getting agreement on this blocks being able to move forward on authn/z.

liggitt commented on docs/design/enhance-pluggable-policy.md:
```diff
@@ -0,0 +1,186 @@
+        `/api/v1beta1/pods`.  For miscelaneous endpoints, like `/version`, the kind is the empty string.
+  - the namespace of the object being access, or the empty string if the endpoint does not support namespaced objects.  This is included in the Context.
+  - content - the runtime.Object being submitted. This only has content for creates and updates and is the object prior to any admission control or defaulting.
+  - resourceName - the name of the resource during a get, update, or delete action.
```
need to update with apiGroup and subresource (or clarify if subresource is included in resource as `resource/subresource`)

ncdc commented:
@erictune have you had a chance to review this yet?

erictune commented:
Okay, taking a look at this now.  Thanks for waiting.

erictune commented:
*Add full Verb to authorizer attributes*: 
Yes

*Add resource name to authorizer attributes*: 
Yes.

*Add content to authorizer attributes*: 
Need more convincing on this.    Reasons I do not like it:

- we may want to support .so authorizer plugins, and I would prefer to keep any .so plugin interface narrow so it does not restrict our ability to change implementation 
- the other attributes can be inferred from headers and path, and so authorization can be done early on, before the body is parsed, converted, etc.  But including content forces us to parse inside the authorization plugin (I don't think that code belongs in a .so pluing) or to parse before authorization (which I think is not ideal).

Can you give more use cases for content or defer adding it?


*Subject access review*: 
Yes, but... I want to clarify some things about it:  
- Is `/policy/{version}/ns/{namespace}/subjectAccessReview` considered an API request or a non-API request?  I think maybe it is a non-API request, but what do you think?   It is different from API requests in that POSTing the request does not create any state on the APIserver, so we sidestep questions about spec and status raised earlier on this thread.  
- Does the namespace in the path do anything?  If not, maybe drop it?
Now we have API groups, and I think we want OpenShift policy objects (roleBinding, etc) to live in a separate API group.  Would you want to put the path under an API group (e.g. /apis/osauthz/v1/ns/foo/subjectAccessReview), or leave it outside any API group.  I think maybe outside might be better, but what do you think? Outside means that it is at a consistent location even if people do not install the OpenShift Authorization API group.  
- Does Subject Access Review just do the kind of authorization that you can do against openshift policy objects (e.g. Role, Policy, and RoleBinding) or does it also check against admission controllers that might reject based on specific content of pods?  I'd lean towards just doing the first category of authorization.
I may have more comments on the details once we sort these things out.

*ResourceAccessReview*: 
yes, but same questions apply as for SubjectAccessReview.

deads2k commented:
> Is /policy/{version}/ns/{namespace}/subjectAccessReview considered an API request or a non-API request? I think maybe it is a non-API request, but what do you think? It is different from API requests in that POSTing the request does not create any state on the APIserver, so we sidestep questions about spec and status raised earlier on this thread.

I think this is an API.  It takes an API type, we want to register it in the API server, it should live in an API group, and we want to protect it with authorization as an API.  Making it an API also makes it very easy and consistent to deal with.

deads2k commented:
> Does the namespace in the path do anything?

Namespace qualification makes it easy to grant permissions to a user related to given namespace.  This is useful in multi-tenant environments.

In openshift, we implemented this with a resource called `LocalSubjectAccessReview`, an admin for a project is granted permission for this endpoint and the impl escalates and calls to the cluster-level SAR with the correct namespace specified.

It makes it very easy to reason about exactly what power you're granting to a user.

deads2k commented:
> Now we have API groups, and I think we want OpenShift policy objects (roleBinding, etc) to live in a separate API group.

I think I'd like to see SAR and RAR under an `authorization.kubernetes.io` group and have things like Roles and RoleBindings under a different group.  The SAR and RAR should be backed by the pluggable authorization engine and the Roles/RoleBindings should be scoped to particular impl of that authorization engine.

deads2k commented:
> Does Subject Access Review just do the kind of authorization that you can do against openshift policy objects (e.g. Role, Policy, and RoleBinding) or does it also check against admission controllers that might reject based on specific content of pods?

It only does the authorization check, the `.Authorize` call.  It doesn't check any admission controllers or the like.

deads2k commented:
> Can you give more use cases for content or defer adding it?

This has only been used a couple spots for handling self-SAR checks.  I think we could replicate the functionality we need by creating a subresource for "self".  I'll have to think about how we could express that.

erictune commented:
Please explain more about the "self-SAR checks" and how the request body is involved.

erictune commented:
> It only does the authorization check, the .Authorize call. It doesn't check any admission controllers or 
> the like.
:+1:

erictune commented:
> I think I'd like to see SAR and RAR under an authorization.kubernetes.io group and have things like Roles and RoleBindings under a different group. The SAR and RAR should be backed by the pluggable authorization engine and the Roles/RoleBindings should be scoped to particular impl of that authorization engine.

Agree with this.

erictune commented:
What does the URL for subject access review look like given the current Kubernetes style for api groups?  Is it: ``/apis/authorization.kubernetes.io/{version}/ns/{namespace}/subjectAccessReview`?

erictune commented:
What is `requestAttributeGetter` going to return for a SAR?
something like:
```
&authorizer.AttributesRecord{
				Verb: "create",
 				Path:            "/apis/extensions/v1beta1/namespaces/myns/jobs",
 				ResourceRequest: true,
 				APIGroup:        "authorization.kubernetes.io",
 				Namespace:       "myns",
 				Resource:        "subjectAccessReview",
			},
```
Do you think it is wierd that the Verb is "create" for a request that does not have any side effects?

deads2k commented:
> Do you think it is wierd that the Verb is "create" for a request that does not have any side effects?

We need the ability to submit data along with the requst, so POST came to mind.  Also, if we go with an SAR that has a `spec` for describing what you want to test and `status` filled in by the server for describing the result, then the "create" metaphor is actually pretty close.  Its not persistent (you can't GET it), but you're submitting data to the API, which is processing it and returning back the "actual" in response to your POST.

smarterclayton commented:
From restful patterns POST is generally the generic action verb.  PUT and
DELETE have very special semantics, POST is "do something on the server
that may change state or not result in a resource being returned".  POST
resource is the backbone of the catch all path for REST.

On Nov 24, 2015, at 3:24 PM, David Eads <notifications@github.com> wrote:

Do you think it is wierd that the Verb is "create" for a request that does
not have any side effects?

We need the ability to submit data along with the requst, so POST came to
mind. Also, if we go with an SAR that has a spec for describing what you
want to test and status filled in by the server for describing the result,
then the "create" metaphor is actually pretty close. Its not persistent
(you can't GET it), but you're submitting data to the API, which is
processing it and returning back the "actual" in response to your POST.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/12209#issuecomment-159395019>
.

erictune commented:
It may be normal for restful patterns in general, but it is the first instance in our API of a POST that does not create something you can GET back, right?

deads2k commented:
I've updated the available attributes to remove `content`.  This forces us to have a `PersonalSubjectAccessReview`.  I have updated the Access Review sections to reflect the groups and the different options since a single Kind cannot be used for multiple Resources and a single Kind cannot be both namespaced and non-namespaced.

erictune commented:
> I think this is an API. It takes an API type, we want to register it in the API server, it should live in an API group, and we want to protect it with authorization as an API. Making it an API also makes it very easy and consistent to deal with.

Just thinking out loud here: 

- We have DeleteOptions which is defined as an API type, and lives in an API group, but which is passed via query parameters rather than as a body.  We could use query parameters from a API type on a non-resource Path.
- nonResourcePaths can be authorized as well, as in #16148.

deads2k commented:
> Just thinking out loud here:
> 
> We have DeleteOptions which is defined as an API type, and lives in an API group, but which is passed via query parameters rather than as a body. We could use query parameters from a API type on a non-resource Path.
> nonResourcePaths can be authorized as well, as in #16148.

I think that we'll end up happier with API resources.  Using resources makes concepts like namespace level protection for the LocalSAR very easy to reason about and inspect in all the authorizers I've seen so far (abac file, keystone, o.  When writing rules for an authorizer, the structure provided by the API resources and divisions provided by namespaces are very easy for users to understand.  Attempting to write similar non-resource URLs rules would be more difficult.

smarterclayton commented:
Query parameters are usually something you use to change a query.  Having a
formal body makes it easier to generate client side structures, but we can
always support a variant in the future that is a HEAD or GET.  Either way,
we what a versioned response.

On Nov 25, 2015, at 9:57 AM, David Eads <notifications@github.com> wrote:

Just thinking out loud here:

We have DeleteOptions which is defined as an API type, and lives in an API
group, but which is passed via query parameters rather than as a body. We
could use query parameters from a API type on a non-resource Path.
nonResourcePaths can be authorized as well, as in #16148
<https://github.com/kubernetes/kubernetes/pull/16148>.

I think that we'll end up happier with API resources. Using resources makes
concepts like namespace level protection for the LocalSAR very easy to
reason about and inspect in all the authorizers I've seen so far (abac
file, keystone, o. When writing rules for an authorizer, the structure
provided by the API resources and divisions provided by namespaces are very
easy for users to understand. Attempting to write similar non-resource URLs
rules would be more difficult.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/12209#issuecomment-159632288>
.

erictune commented:
I talked to @bgrant0607.  He doesn't have a problem in principle with the  but would like some time to think about the conventions for such a resource.

Can you proceed with your implementation  of `/apis/authorization.kubernetes.io/{version}/ns/{namespace}/subjectAccessReview` with version `v1beta1` but be open to changing the naming or path structure before going to `v1`?

deads2k commented:
> Can you proceed with your implementation of /apis/authorization.kubernetes.io/{version}/ns/{namespace}/subjectAccessReview with version v1beta1 but be open to changing the naming or path structure before going to v1?

Yes.  As soon as API Groups seem stable enough to build on top of (see https://github.com/kubernetes/kubernetes/issues/16173 and https://github.com/kubernetes/kubernetes/issues/17216 for ongoing work), I'll add it.

erictune commented:
LGTM

wojtek-t commented:
@deads2k 

FAIL: changes needed but not made due to --verify
/home/travis/gopath/src/github.com/kubernetes/kubernetes/docs/ is out of date. Please run hack/update-generated-docs.sh

Can you please fix it?
